### PR TITLE
Check array floats

### LINKED ISSUE #871: reshape incorrectly handles floats
Related to #758,

The following code behaves as expected:
```
import dask.array as da
import numpy as np

data = np.arange(int(1e6))
dask_array = da.from_array(data, chunks=(100, ))

window_size = float(100.1)
number_of_windows = 10

sliced = dask_array[:window_size*number_of_windows].reshape((window_size, number_of_windows)).compute()

#####

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-29-ae89621fb406> in <module>()
      8 number_of_windows = 10
      9 
---> 10 sliced = dask_array[:window_size*number_of_windows].reshape((window_size, number_of_windows)).compute()

/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py in reshape(self, *shape)
    888         if len(shape) == 1 and not isinstance(shape[0], Number):
    889             shape = shape[0]
--> 890         return reshape(self, shape)
    891 
    892     @wraps(topk)

/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py in reshape(array, shape)
   2222     if np.prod(shape) != array.size:
   2223         raise ValueError('total size of new array must be unchanged')
-> 2224     return unravel(ravel(array), shape)
   2225 
   2226 

/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py in unravel(array, shape)
   2207             return stack([unravel(array[n * trailing_size
   2208                                         : (n + 1) * trailing_size], shape[1:])
-> 2209                           for n in range(shape[0])])
   2210 
   2211 

TypeError: 'float' object cannot be interpreted as an integer
``` 

But using -1 to infer the float dimension gives a less helpful error:

```

import dask.array as da
import numpy as np

data = np.arange(int(1e6))
dask_array = da.from_array(data, chunks=(100, ))

window_size = float(100.1)
number_of_windows = 10

sliced = dask_array[:window_size*number_of_windows].reshape((-1, number_of_windows)).compute()

######

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-28-44273dd5d7c1> in <module>()
      8 number_of_windows = 10
      9 
---> 10 sliced = dask_array[:window_size*number_of_windows].reshape((-1, number_of_windows)).compute()

/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py in reshape(self, *shape)
    888         if len(shape) == 1 and not isinstance(shape[0], Number):
    889             shape = shape[0]
--> 890         return reshape(self, shape)
    891 
    892     @wraps(topk)

/Users/jeremy/anaconda/lib/python3.5/site-packages/dask/array/core.py in reshape(array, shape)
   2221 
   2222     if np.prod(shape) != array.size:
-> 2223         raise ValueError('total size of new array must be unchanged')
   2224     return unravel(ravel(array), shape)
   2225 

ValueError: total size of new array must be unchanged

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #871 

This disallows float slices and shapes with non-zero mantissas.

### CODE REVIEW

File: dask/array/slicing.py [hunk 1]
```diff
@@ -11,27 +12,47 @@
...
+    >>> sanitize_index(0.5)
+    Traceback (most recent call last):
+    ...
+    IndexError: Bad index.  Must be integer-like: 0.5
     """
-    if isinstance(ind, np.integer):
-        ind = int(ind)
+    if isinstance(ind, Number):
+        ind2 = int(ind)
+        if ind2 != ind:
```
Comment (shoyer) [inline dask/array/slicing.py]: Arguably, you should use `operator.index` instead -- it exists for exactly this purpose.

**Comment by mrocklin** [comment]
cc @jeremyherbert

File: dask/array/slicing.py (hunk 1 above)
Comment (mrocklin) [inline dask/array/slicing.py]: This doesn't allow through values like `1.0` which is useful for lines like the following in `reshape`

    missing_size = sanitize_index(array.size / np.prod(known_sizes))

Comment (mrocklin) [inline dask/array/slicing.py]: We could move this logic there.  I'm tempted to allow integer-equivalent floats though.

**Pull request merged**